content/blog/tidymodels-errors-q4/index.Rmd

```diff
@@ -0,0 +1,229 @@
+
+mtcars_rec_trained %>% bake(NULL)
+```
+
+Mmm. What happened to `mpg`? We mistakenly told `step_pca()` to perform PCA on _all_ of the numeric variables, not just the numeric _predictors_! As a result, it incorporated `mpg` into the principal components, removing each of the original numeric variables after the fact. Rewriting using the correct tidyselect specification `all_numeric_predictors()`:
+
+```{r}
+mtcars_rec_new <- 
+  recipe(mpg ~ ., mtcars) %>%
+  step_pca(all_numeric_predictors())
+
+workflow(mtcars_rec_new, linear_spec) %>% fit(mtcars)
+```
+
+Works like a charm. That error we saw previously could be much more helpful, though. With the current developmental version of parsnip, this looks like:
+
+```{r no-outcome-parsnip, error = TRUE}
+fit(linear_spec, ~ hp, mtcars)
+```
+
+Or, with workflows:
+
+```{r no-outcome-workflows, error = TRUE}
+workflow(mtcars_rec, linear_spec) %>% fit(mtcars)
+```
+
+Much better.
+
+## Predicting with things that can't predict {#predict}
+
+Earlier this year, Dr. Louise E. Sinks put out a [wonderful blog post](https://lsinks.github.io/posts/2023-04-10-tidymodels/tidymodels_tutorial.html) documenting what it felt like to approach the various object types defined in the tidymodels as a newcomer to the collection of packages. They wrote:
+
+> I found it confusing that `fit`, `last_fit`, `fit_resamples`, etc., did not all produce objects that contained the same information and could be acted on by the same functions. 
+
+This makes sense. While we try to forefront the intended mental model for fitting and predicting with tidymodels in our APIs and documentation, we also needed to be proactive in anticipating common lapses in constructing that mental model.
+
+For example, we've found that it's sometimes not clear to users which outputs they can call `predict()` on. One such situation, as Louise points out, is with `fit_resamples()`:
+
+```{r fit-resamples}
+# fit a linear regression model to bootstrap resamples of mtcars
+mtcars_res <- fit_resamples(linear_reg(), mpg ~ ., bootstraps(mtcars))
+
+mtcars_res
+```
+
+With previous tidymodels versions, mistakenly trying to predict with this object resulted in the following output:
+
+``` r
+predict(mtcars_res)
+#> Error in UseMethod("predict") : 
+#>   no applicable method for 'predict' applied to an object of class
+#>   "c('resample_results', 'tune_results', 'tbl_df', 'tbl', 'data.frame')"
+```
+
+Some R developers may recognize this error as what results when we didn't define any `predict()` method for `tune_results` objects. We didn't do so because prediction isn't well-defined for tuning results. _But_, this error message does little to help a user understand why that's the case.
+
+We've recently made some changes to error more informatively in this case. We do so by defining a "dummy" `predict()` method for tuning results, implemented only for the sake of erroring more informatively. The same code will now give the following output:
+
+```r
```

Comment:
In other cases, I've only used `r` in favor of `{r}` for output from older (current CRAN) package versions. I do so here since the inline `{.help}` markup can be difficult to read in markdown output.